pkg/registry/etcd/ns_server.go

```diff
@@ -73,24 +74,34 @@ func (n *etcdNSRegistryServer) Register(ctx context.Context, request *registry.N
 }
 
 func (n *etcdNSRegistryServer) watch(query *registry.NetworkServiceQuery, s registry.NetworkServiceRegistry_FindServer) error {
-	watcher, err := n.client.NetworkservicemeshV1().NetworkServices(n.ns).Watch(s.Context(), metav1.ListOptions{})
-	if err != nil {
-		return err
-	}
+	logger := log.FromContext(n.chainContext).WithField("etcdNSRegistryServer", "watch")
+
 	for {
-		select {
-		case <-s.Context().Done():
-			return s.Context().Err()
-		case event := <-watcher.ResultChan():
-			if event.Type != watch.Added {
-				continue
-			}
-			model := event.Object.(*v1.NetworkService)
-			item := (*registry.NetworkService)(&model.Spec)
-			if matchutils.MatchNetworkServices(query.NetworkService, item) {
-				err := s.Send(item)
-				if err != nil {
-					return err
+		watcher, err := n.client.NetworkservicemeshV1().NetworkServices(n.ns).Watch(s.Context(), metav1.ListOptions{})
+		if err != nil {
+			return err
+		}
```

Comment:
Why do we need to re-create watcher?

Reply:
We are still processing watching Find, so we still need a watcher.

Reply:
As you internally pointed out, we don't actually have the option to use a perpetual observer https://github.com/kubernetes/apiserver/blob/master/pkg/endpoints/handlers/get.go#L258